File: Difficulty: Medium/Combination Sum II/combination-sum-ii.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.util.*;

class GFG {
    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        int t;
        t = sc.nextInt();
        while (t-- > 0) {

            int n = sc.nextInt();
            int k = sc.nextInt();
            int[] candidates = new int[n];
            for (int i = 0; i < n; i++) candidates[i] = sc.nextInt();

            Solution obj = new Solution();
            List<List<Integer>> ans = obj.combinationSum2(candidates, k);
            Collections.sort(ans, new Comparator<List<Integer>>() {
                @Override   public int compare(List<Integer> a, List<Integer> b) {
                    int al = a.size();
                    int bl = b.size();
                    int min = Math.min(al, bl);
                    for (int i = 0; i < min; i++) {
                        if (a.get(i) < b.get(i))
                            return -1;
                        else if (a.get(i) > b.get(i))
                            return 1;
                    }
                    if (al < bl)
                        return -1;
                    else if (al > bl)
                        return 1;
                    return -1;
                }
            });
            System.out.print("[ ");
            for (int i = 0; i < ans.size(); i++) {
                System.out.print("[ ");
                for (int j = 0; j < ans.get(i).size(); j++)
                    System.out.print(ans.get(i).get(j) + " ");
                System.out.print("]");
            }
            System.out.println(" ]");
        
System.out.println("~");
}
    }
}
// } Driver Code Ends

class Solution {
    
    /**
     * Helper function to generate combinations that sum up to the target `k`
     * @param ind Starting index for the current recursion
     * @param ds A temporary list to store the current combination
     * @param a The input array
     * @param n The length of the array
     * @param k The remaining sum to find
     * @param res The result list of all combinations
     */
    static void combinationsum2(int ind, List<Integer> ds, int[] a, int n, int k, List<List<Integer>> res) {
        // If the remaining sum is 0, add the current combination to the result
        if(k == 0){
            res.add(new ArrayList<>(ds));
            return;
        }
        
        for(int i = ind; i < n; i++) {
            // Skip duplicate elements to avoid repeated combinations
            if(i > ind && a[i] == a[i - 1]) continue;
            
            // If the current element is less than or equal to the remaining sum
            if(a[i] <= k){
                // Add the element to the current combination
                ds.add(a[i]);
                // Recur with reduced sum and next index
                combinationsum2(i + 1, ds, a, n, k - a[i], res);
                // Backtrack by removing the last added element
                ds.remove(ds.size() - 1);
            }
        }
    }
    
    /**
     * Function to find all unique combinations in the array `a` that sum up to `s`
     * @param a The input array of integers
     * @param s The target sum
     * @returns List of unique combinations that sum up to `s`
     */
    public static List<List<Integer>> combinationSum2(int[] a, int s) {
        Arrays.sort(a);  // Sort the array to handle duplicates
        List<List<Integer>> res = new ArrayList<>();  // Result list to store the combinations
        List<Integer> ds = new ArrayList<>();  // Temporary list to build the current combination
        int n = a.length;
        
        // Start the recursive function with the initial values
        combinationsum2(0, ds, a, n, s, res);
        
        return res;  // Return the result
    }
}

File: Difficulty: Medium/Combination Sum II/combination-sum-ii.js
//{ Driver Code Starts

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => { inputString += inputStdin; });

process.stdin.on('end', _ => {
    inputString =
        inputString.trim().split("\n").map(string => { return string.trim(); });

    main();
});

function readLine() { return inputString[currentLine++]; }

function main() {
    let t = parseInt(readLine());
    let i = 0;
    for (; i < t; i++) {

        let coins = readLine().trim().split(" ").map((x) => parseInt(x));
        let sum = parseInt(readLine());
        let obj = new Solution();
        let res = obj.uniqueCombinations(coins, sum);
        if (res.length == 0) {
            console.log(-1);
        } else {
            for (let row of res) {
                row.sort((a, b) => a - b); // Sort each row in ascending order
            }

            // Step 2: Sort `res` lexicographically
            res.sort((row1, row2) => {
                const len = Math.min(row1.length, row2.length);
                for (let i = 0; i < len; i++) {
                    if (row1[i] !== row2[i]) {
                        return row1[i] - row2[i]; // Compare row elements
                    }
                }
                return row1.length -
                       row2.length; // If all elements are equal, sort by row length
            });
            let ans = "";
            for (let row of res) {
                let s = "[";
                for (let k = 0; k < row.length; k++) {
                    s += row[k];
                    s += " ";
                }
                s += "]";
                ans += s;
                ans += " ";
            }
            console.log(ans);
        }
        console.log("~");
    }
}

// } Driver Code Ends

// User function Template for JavaScript

/**
 * @param {number[]} arr
 * @param {number} target
 * @returns {number[][]}
 */

class Solution {
    // Function to find all unique combinations of elements in array 'arr' 
    // that sum up to the 'target' value.
    uniqueCombinations(arr, target) {
       const  resultCombinations = [], currentCombination = []; 
      
        // Sort the array to ensure combinations are considered in ascending order.
        arr.sort((a, b) => a - b);

        // Initialize current sum to track the total of the selected combination.
        let currentSum = 0;
        
        // Variable to store the last element added to the combination, 
        // used to avoid duplicates in the same recursion path.
        let lastElement = 0;

        // Recursive helper function to generate all possible combinations.
        const findCombinations = (index) => {
            // If the current sum matches the target, add the combination to the result.
            if (currentSum === target) return resultCombinations.push([...currentCombination]);

            // If the current sum exceeds the target or we've exhausted all elements, stop recursion.
            if (currentSum > target || index === arr.length) return;

            // Skip the current index if the element is the same as the last one added 
            // to avoid generating duplicate combinations.
            if (lastElement !== arr[index]) {
                // Add the current element to the combination and update the current sum.
                currentCombination.push(arr[index]);
                currentSum += arr[index];

                // Recurse with the next index to explore combinations including the current element.
                findCombinations(index + 1);

                // If the current sum exceeds the target, backtrack by removing the last element added.
                if (currentSum > target) return currentCombination.pop(), currentSum -= arr[index];

                // Backtrack: Remove the current element and adjust the current sum.
                currentCombination.pop();
                currentSum -= arr[index];

                // Update lastElement to prevent reusing the same element in this recursion path.
                lastElement = arr[index];
            }

            // Recurse without including the current element to explore alternative combinations.
            findCombinations(index + 1);
        };

        // Start the recursion with the first element in the array.
        findCombinations(0);

        // Return the list of all unique combinations that sum up to the target.
        return resultCombinations;
    }
}
